fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/LogicalProperty.java

```diff
@@ -66,21 +80,56 @@ public OneTabletProperty oneTabletProperty() {
 
     public LogicalProperty() {
         this.outputColumns = new ColumnRefSet();
+        this.usedCTEs = CTEProperty.EMPTY;
     }
 
     public LogicalProperty(ColumnRefSet outputColumns) {
         this.outputColumns = outputColumns;
+        this.usedCTEs = CTEProperty.EMPTY;
     }
 
     public LogicalProperty(LogicalProperty other) {
         outputColumns = other.outputColumns.clone();
         oneTabletProperty = other.oneTabletProperty;
+        usedCTEs = other.usedCTEs;
     }
 
     public void derive(ExpressionContext expressionContext) {
         LogicalOperator op = (LogicalOperator) expressionContext.getOp();
         outputColumns = op.getOutputColumns(expressionContext);
         oneTabletProperty = op.accept(new OneTabletExecutorVisitor(), expressionContext);
+        if (expressionContext.isGroupExprContext()) {
+            // only derived after entering memo
+            deriveUsedCTEs(expressionContext);
+        }
+    }
+
+    private void deriveUsedCTEs(ExpressionContext expressionContext) {
+        OperatorType type = expressionContext.getOp().getOpType();
+        Set<Integer> cteIds = Sets.newHashSet();
+
+        if (type == LOGICAL_CTE_ANCHOR) {
+            LogicalCTEAnchorOperator anchorOperator = (LogicalCTEAnchorOperator) expressionContext.getOp();
+            cteIds.addAll(expressionContext.getChildLogicalProperty(0).getUsedCTEs().getCteIds());
+            cteIds.addAll(expressionContext.getChildLogicalProperty(1).getUsedCTEs().getCteIds());
+            cteIds.remove(anchorOperator.getCteId());
+        } else if (type == LOGICAL_CTE_PRODUCE) {
+            LogicalCTEProduceOperator produceOperator = (LogicalCTEProduceOperator) expressionContext.getOp();
+            cteIds.addAll(expressionContext.getChildLogicalProperty(0).getUsedCTEs().getCteIds());
+            cteIds.remove(produceOperator.getCteId());
```

Comment:
the children of producer never support producer's cteid

Reply:
Yes, the next remove step is unnecessary.